pycqed/measurement/detector_functions.py

```diff
@@ -42,6 +43,41 @@ def prepare(self, **kw):
     def finish(self, **kw):
         pass
 
+    def generate_metadata(self):
+        det_metadata = {k: self.savable_attribute_value(v, self.name)
+                        for k, v in self.__dict__.items()}
+
+        detectors_dict = {}
+        for d in det_metadata.pop('detectors', []):
+            detectors_dict.update({f'{d["UHFs"][0]} {d["name"]}': d})
+        if len(detectors_dict):
+            det_metadata['detectors'] = detectors_dict
+
+        return {'Detector Metadata': det_metadata}
+
+    @staticmethod
+    def savable_attribute_value(attr_val, det_name):
```

Comment:
I think you should add a comment here about what this function does, since it is in the base class and used by all detector functions.

I am not exactly sure if this function couldn't be encorporated into `write_dict_to_hdf5`. If it can, then it would be nicer there and could be reused also by the analysis code for example.

Reply:
Thanks for looking at it!
I'm not sure I understand what you mean that "then it could be reused by the analysis code." The new Detector Metadata group in is i the Experimental Metadata outer group so it will automatically exist inside the self.raw_data_dict['exp_metadata'].

Reply:
Thanks for the comments!
You mean if the `savable_attribute_value` is in `write_data_to_hdf5` it's nicer because the analysis can use it when it saves to the hdf? But analysis never has access to a detector function class instance, which is where this function is relevant.  Nor does the analysis ever save anything about detector functions.
`savable_attribute_value` is a helper function specifically for detector function. It makes sure that if any of the det_dunct attributes are class instances (like det_func.AWG), they are passed to the metadata as class_instance.name instead of class_instance in which case it would be saves as a string "<Pulsar: Pulsar>".

Reply:
I guess what I had in mind in that case is that we could store the detector function itself in the metadata, and then when the MC saves the metadata (using write_dict_to_hdf5), the detector function (or some other class' parameters if we want to store other objects in the metadata) attributes get converted to the suitable format there. That way the code that does the same function is located in the same place and can more easily be reused.

Handling the nested detector functions might be more difficult in my proposed way, though.
I think that you can merge the additions as they are definitely useful, but just keep the comments in mind when reworking the metadata storage at some point in the future. We can discuss on zoom if something was still unclear to me or to you.

Reply:
OK I understand what you had in mind. It could work if we instead define a standard method like "get_metadata" that the function write_dict_to_hdf5 tries to call whenever it finds a class instance. For now only the detector functions will have it but he functionality could be extended in the future. This would also be a relatively easy change. What do you think?